pkg/controller/persistentvolumelabel/pvlcontroller.go

```diff
@@ -0,0 +1,368 @@
+
+	"k8s.io/client-go/kubernetes"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+
+	"k8s.io/kubernetes/pkg/cloudprovider"
+	"k8s.io/kubernetes/pkg/cloudprovider/providers/aws"
+	"k8s.io/kubernetes/pkg/cloudprovider/providers/gce"
+	"k8s.io/kubernetes/pkg/controller"
+	kubeletapis "k8s.io/kubernetes/pkg/kubelet/apis"
+	vol "k8s.io/kubernetes/pkg/volume"
+)
+
+const initializerName = "pvlabel.kubernetes.io"
+const initialzerConfigName = "pvlabel-initconfig"
+
+// PersistentVolumeLabelController handles adding labels to persistent volumes when they are created
+type PersistentVolumeLabelController struct {
+	// Control access to cloud volumes
+	mutex            sync.Mutex
+	ebsVolumes       aws.Volumes
+	gceCloudProvider *gce.GCECloud
+
+	cloud         cloudprovider.Interface
+	kubeClient    kubernetes.Interface
+	pvlController cache.Controller
+	volumeLister  corelisters.PersistentVolumeLister
+
+	syncHandler func(vol *v1.PersistentVolume) error
+
+	// queue is where incoming work is placed to de-dup and to allow "easy" rate limited requeues on errors
+	queue workqueue.RateLimitingInterface
+}
+
+// NewPersistentVolumeLabelController creates a PersistentVolumeLabelController object
+func NewPersistentVolumeLabelController(
+	kubeClient kubernetes.Interface,
+	cloud cloudprovider.Interface) *PersistentVolumeLabelController {
+	var pvlIndexer cache.Indexer
+
+	pvlc := &PersistentVolumeLabelController{
+		cloud:      cloud,
+		kubeClient: kubeClient,
+		queue:      workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "pvLabels"),
+	}
+	pvlc.syncHandler = pvlc.AddLabels
+	pvlIndexer, pvlc.pvlController = cache.NewIndexerInformer(
+		&cache.ListWatch{
+			ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+				options.IncludeUninitialized = true
+				return kubeClient.CoreV1().PersistentVolumes().List(options)
+			},
+			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+				options.IncludeUninitialized = true
+				return kubeClient.CoreV1().PersistentVolumes().Watch(options)
+			},
+		},
+		&v1.PersistentVolume{},
+		0,
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				key, err := cache.MetaNamespaceKeyFunc(obj)
+				if err == nil {
+					pvlc.queue.Add(key)
+				}
+			},
+		},
+		cache.Indexers{},
+	)
+	pvlc.volumeLister = corelisters.NewPersistentVolumeLister(pvlIndexer)
+
+	return pvlc
+}
+
+// Run starts a controller that adds labels to persistent volumes
+func (pvlc *PersistentVolumeLabelController) Run(threadiness int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer pvlc.queue.ShutDown()
+
+	glog.Infof("Starting PersistentVolumeLabelController")
+	defer glog.Infof("Shutting down PersistentVolumeLabelController")
+
+	go pvlc.pvlController.Run(stopCh)
+
+	if !controller.WaitForCacheSync("persistent volume label", stopCh, pvlc.pvlController.HasSynced) {
+		return
```

Comment:
`controller.WaitForCacheSync` is sufficient - we wrote it to log that the caches didn't get synced.